src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractGenerateNuGetPackageCommand.cs

```diff
@@ -58,7 +58,7 @@ private async Task<bool> IsReadyToBuildAsync()
             // Ensure build manager is initialized.
             await EnsureBuildManagerInitializedAsync();
 
-            ErrorHandler.ThrowOnFailure(_buildManager.QueryBuildManagerBusy(out int busy));
+            ErrorHandler.ThrowOnFailure(_buildManager!.QueryBuildManagerBusy(out int busy));
```

Comment:
I wish there was a way to restructure this so the compiler could track this better

Reply:
Agreed.  It's getting to be a pain.

Reply:
Such restructuring will likely help formalise object lifecycles that were previously implicit.

On this class it's possible to convert `Task EnsureBuildManagerInitializedAsync()` into `Task<IBuildManager> GetBuildManager()` which always returns a non-null instance to a local variable (and itself is the only method to interact with `_buildManager`), and `IsReadyToBuildAsync()` could be converted to a synchronous extention method.

In more complex cases where initialization populates a lot of state, moving that state into a new class where a `EnsureInitialized`-like method returns a non-null instance might help.

Such changes require a fair bit of thought per instance. It'll be interesting to see what patterns emerge and how they effect code quality.

Reply:
Definitely agree, and have been thinking about the opportunities to extract things into classes together to help enforce things.  See the (simple) change I made in the up to date logging as an example.